packages/expo-in-app-purchases/ios/EXInAppPurchases/EXInAppPurchasesModule.m

```diff
@@ -0,0 +1,449 @@
+        break;
+      }
+    }
+  }
+}
+
+# pragma mark - Handling Promises
+
+// Returns a boolean indicating the promise was successfully set. Otherwise, we should return immediately
+- (BOOL)setPromise:(NSString*)key resolve:(UMPromiseResolveBlock)resolve reject:(UMPromiseRejectBlock)reject
+{
+  NSArray *promise = _promises[key];
+  
+  if (promise == nil) {
+    _promises[key] = @[resolve, reject];
+    return YES;
+  } else {
+    reject(@"E_UNFINISHED_PROMISE", @"Must wait for promise to resolve before recalling function.", nil);
+    return NO;
+  }
+}
+
+- (void)resolvePromise:(NSString*)key value:(id)value
+{
+  NSArray *currentPromise = _promises[key];
+  
+  if (currentPromise != nil) {
+    UMPromiseResolveBlock resolve = currentPromise[0];
+    _promises[key] = nil;
+    
+    resolve(value);
+  }
+}
+
+- (void)rejectPromise:(NSString*)key code:(NSString*)code message:(NSString*)message error:(NSError*) error
+{
+  NSArray* currentPromise = _promises[key];
+  
+  if (currentPromise != nil) {
+    UMPromiseRejectBlock reject = currentPromise[1];
+    _promises[key] = nil;
+    
+    reject(code, message, error);
+  }
+}
+
+# pragma mark - Formatting Response
+
+- (NSDictionary *)getProductData:(SKProduct *)product
+{
+  // Use with caution: P0D also implies non-renewable subscription.
+  NSString *subscriptionPeriod = [self getSubscriptionPeriod:product];
+  NSString *type = [subscriptionPeriod isEqualToString:kEXInAppSubPeriod] ? kEXInApp : kEXSubs;
+  
+  NSDecimalNumber *oneMillion = [[NSDecimalNumber alloc] initWithInt:1000000];
+  NSDecimalNumber *priceAmountMicros = [product.price decimalNumberByMultiplyingBy:oneMillion];
+  NSString *price = [NSString stringWithFormat:@"%@%@", product.priceLocale.currencySymbol, product.price];
+  
+  return @{
+           @"description": product.localizedDescription,
+           @"price": price,
+           @"priceAmountMicros": priceAmountMicros,
+           @"priceCurrencyCode": product.priceLocale.currencyCode,
+           @"productId": product.productIdentifier,
+           @"subscriptionPeriod": subscriptionPeriod,
+           @"title": product.localizedTitle,
+           @"type": type
+           };
+}
+
+- (NSDictionary *)getTransactionData:(SKPaymentTransaction *)transaction acknowledged:(BOOL)acknowledged
+{
+  NSData *receiptData = [NSData dataWithContentsOfURL:[[NSBundle mainBundle] appStoreReceiptURL]];
+  
+  // Get original transaction info if any
+  SKPaymentTransaction *originalTransaction = transaction.originalTransaction;
+  NSString *originalTransactionId = originalTransaction ? originalTransaction.transactionIdentifier : NSString.string;
+  NSNumber *originalTransactionTime = originalTransaction ?
+  @(originalTransaction.transactionDate.timeIntervalSince1970 * 1000) : @(0);
+  
+  return @{
+           @"acknowledged": @(acknowledged),
+           @"productId": transaction.payment.productIdentifier,
+           @"purchaseToken": transaction.transactionIdentifier,
```

Comment:
Same note as the one below - this is not a token, but a simple numeric identifier, similar to android's `orderId` property (though not the same in behaviour)

Reply:
I know the naming here is a bit confusing. I was originally returning the transaction ID as the "orderID" since they are very similar. My goal here is to reduce the number of platform specific return values and provide abstraction wherever possible.

The reasoning behind renaming it "purchaseToken" despite being a bit more opaque was so that `finishTransactionAsync` would take in the same parameter. I wanted to avoid a situation like this:
```javascript
if (!purchase.acknowledged) {
  finishTransactionAsync(Platform.OS === 'ios' ? purchase.transactionID : purchase.purchaseToken);
}
```
Apple defines transaction ID as:
> A string that uniquely identifies a successful payment transaction.

Similarly, Google defines a purchase token as:
> A token that uniquely identifies a purchase for a given item and user pair. 

and an order ID as:
> A unique order identifier for the transaction.

The key here is "uniquely identifies" which is where I felt abstraction was best since transactionIDs and purchase tokens are closely related. Obviously any discrepancies between native and JS naming will be documented, as long as all the important stuff is returned.

Reply:
Resolved in https://github.com/expo/expo/pull/4416/commits/d3cf1a0776fd0438d879d301f56294d40c0e50e1